pytest_playwright/pytest_playwright.py

```diff
@@ -419,3 +442,112 @@ def pytest_addoption(parser: Any) -> None:
         default=False,
         help="Whether to take a full page screenshot",
     )
+
+
+class ArtifactsRecorder:
+    def __init__(
+        self, pytestconfig: Any, request: pytest.FixtureRequest, playwright: Playwright
+    ) -> None:
+        self._request = request
+        self._pytestconfig = pytestconfig
+        self._playwright = playwright
+
+        self._open_contexts: BrowserContext = []
+        self._all_pages: List[Page] = []
+        self._traces: List[str] = []
+        self._tracing_option = pytestconfig.getoption("--tracing")
+        self._capture_trace = self._tracing_option in ["on", "retain-on-failure"]
+
+    def did_finish_test(self, failed: bool) -> None:
+        screenshot_option = self._pytestconfig.getoption("--screenshot")
+        capture_screenshot = screenshot_option == "on" or (
+            failed and screenshot_option == "only-on-failure"
+        )
+        if capture_screenshot:
+            for index, page in enumerate(self._all_pages):
+                human_readable_status = "failed" if failed else "finished"
+                screenshot_path = _build_artifact_test_folder(
+                    self._pytestconfig,
+                    self._request,
+                    f"test-{human_readable_status}-{index+1}.png",
+                )
+                try:
+                    page.screenshot(
+                        timeout=5000,
+                        path=screenshot_path,
+                        full_page=self._pytestconfig.getoption(
+                            "--full-page-screenshot"
+                        ),
+                    )
+                except Error:
+                    pass
+
+        # Close contexts which were not closed during the test (this will trigger Trace and Video generation)
+        while len(self._open_contexts) > 0:
+            self._open_contexts[0].close()
+
+        if self._tracing_option == "on" or (
+            failed and self._tracing_option == "retain-on-failure"
+        ):
+            for index, trace in enumerate(self._traces):
+                retain_trace = self._capture_trace or failed
+                trace_file_name = (
+                    "trace.zip" if len(self._traces) == 1 else f"trace-{index+1}.zip"
+                )
+                trace_path = _build_artifact_test_folder(
+                    self._pytestconfig, self._request, trace_file_name
+                )
+                if retain_trace:
+                    os.makedirs(os.path.dirname(trace_path), exist_ok=True)
+                    shutil.move(trace, trace_path)
+                else:
+                    os.remove(trace)
+
+        video_option = self._pytestconfig.getoption("--video")
+        preserve_video = video_option == "on" or (
+            failed and video_option == "retain-on-failure"
+        )
+        if preserve_video:
+            for index, page in enumerate(self._all_pages):
```

Comment:
Let's see whether we can delete unnecessary videos as well.

Reply:
Addressed and verified.